openvdb_ax/openvdb_ax/compiler/Logger.cc

```diff
@@ -187,19 +187,19 @@ void Logger::setSourceCode(const char* code)
 bool Logger::error(const std::string& message,
                    const Logger::CodeLocation& lineCol)
 {
-    // already exceeded the error limit
-    if (this->atErrorLimit()) return false;
+    // check if we've already exceeded the error limit
+    const bool limit = this->atErrorLimit();
+    // Always increment the error counter
+    ++mNumErrors;
+    if (limit) return false;
```

Comment:
Isnt the whole point of max erroring to limit the error counter? If i now query this for some print out in my application it will say one more than my max error?

Reply:
This is kind of the point of the max errors limit though. I think it is weirder to have the ability to have errors() that doesnt match the number of entries that have been logged. errors() is there to be the way to determine how many things have been logged, e.g. if I'm printing messages to std::cerr and I want to say afterwards say ''x' errors found'. Once the Logger has hit maxErrors it shouldnt carry on with compilation, and should exit.

Reply:
Okay, then the logger can't be reliably used for error termination and should never have been - each individual method will need to track this alongside the logger.

Reply:
I was unaware that any syntax error from the parser could occur and still return an AST, so when it goes into the compiler the assumption was if it hadn't errored yet, you would be working with a legit AST. I think the only other place where this was actually being the sole marker for erroring was in verifyTypedAccesses however I think a failure there will NOT stop a compile so that actually doesn't need the change above. eg. I might actually want my snippet to try and compile if the typed accesses are different, and get an error saying the type is wrong in various places.

Reply:
What if you use a logger which previously errored in compialtion in a new call to `parse`? This is totally valid, but now the logger has an error state. This paradigm dosesn't work. Checking `hasError()` is not reliable - the object could literally have any state. If the class cannot continue to track errors it cannot be used for reliable termination unless you add some other infrastructure - which seems excesive.

Reply:
If this isn't clear, consider the following:
```
Logger lg.
lg.setMaxErrors(1);
auto ast = parse(tree);
if (usesFunction(ast, "bad")) lg.error("you shouldn't use this function");
compile(ast, lg); // what do you want this to do - it **needs** to be able to compile and track if COMPILATION errored
```

Reply:
Because the logger has some non trivial state e.g. its AST node to Code Location map, it relies on users calling clear() between parse calls, in fact parse() should probably do that (I'm not sure why it doesnt). I would imagine in this scenario it should check if the maxErrors has been hit and then subsequently call or not call compile? then max errors is doing its thing of allowing early exit.

Reply:
Discussed offline, we should allow the counter to increase beyond maxerrors but add a comment stating how to get the number of actually reported errors in Logger.h